### PR TITLE
Correctness fixes, TRANSFORM region sizing, a Python-model API, a perf gate

### PR DESCRIPTION
**Pull request by vboussot**
Eighteen commits, one per area. Every suite is green on the tip: `pixi run check` (core + apps), mypy on the four packages, the konfai-mcp and Studio suites, the eleven Playwright tests, the strict docs build.

## What changes

- **Core** (`fix(core)`): refusals and counts (no-optimizer TRAIN, masked selection, unsigned Dice, weightless or EMA-only checkpoints, Trim on a padded patch, 2.5D context on the region route, owned OME-Zarr cache entries); scheduler and scaler state in the checkpoint with occurrence identities; one DDP context over forward and backward with deferred nested steps; RESUME from a versioned epoch-boundary cursor with `resume_latest.pt`; a 1 GiB checkpoint cache (`Predictor.checkpoint_cache_gib`) with memory-mapped loads; the pinned-buffer borrow on offload (64 MiB fp16 patch 17.3 to 7.2 ms); nnU-Net weights through `pretrained_from`.
- **Config**: chains spelled as YAML lists under occurrence keys (`Clip`, `Clip#2`), one scalar coercion on every binding path including typed containers and unions.
- **TRANSFORM**: regions sized by price for the first one and by what the last held for the next, LOAD decided by the store's read capability, halve-on-OOM on CUDA. ExaSPIM (513x1331x1775 uint16): 1G 24.5 to 18.8 s, 2G 17.5 to 11.2, 16G 7 to 5.5; the 2 GiB h5 transform under a 1 GiB budget 7.4 to 3.9 s.
- **API**: `konfai.train_model` / `konfai.predict_model` on any `nn.Module`, `import_bundle` / `export_bundle` for MONAI Bundles, two examples (BringYourModel, LargeImages on the public ExaSPIM store).
- **Apps, MCP, Studio**: streamed multipart uploads (requests-toolbelt), cancellation and lease races, bundle replacement kept whole, portable roles, owned scratch copies for validation, bounded log feeds, lazy panels, browser tests (Playwright).
- **Perf and CI**: `benchmarks/perf` with a committed baseline and `pixi run perf-check` (exit 1 on a regression, 2 on invalid evidence), a nightly workflow for a self-hosted runner, representative PR lanes, mypy in the hook and in CI, one torch thread per test worker (the suite 8x faster on 24 cores).

Two new dependencies, both agreed: `requests-toolbelt` (konfai-apps) and `@playwright/test` (Studio dev). The docs site reduction is a separate PR stacked on this one.

## After the review

Five more commits: checkpoint selection scores what each loss minimizes (a Dice reporting its coefficient summed with a cross-entropy kept the worst epoch), bundle support files declared at export, a two-class CPU segmentation quickstart (twenty epochs in about seven seconds, Dice 0.98), the CI-only platform fixes (Windows, no CUDA, torch 2.14 typing, the pre-commit job), and the CodeRabbit findings (live-model token released, tuple annotations, HDF5 grain of a recovered entry, MCP snapshot on the terminal transition, the bench scripts). One thread stays open: the perf baseline is recaptured from a clean checkout on a quiet machine before the tag.

## Measured, not claimed

Training speed against 1.4.9 on the same torch and GPU: 1.8.3 is 3.6x faster per epoch on the 2D Segmentation example and 1.65x on a synthetic 3D UNet; the current defaults (autocast, channels_last, pin_memory) 7.6x and 4.5x.

## Summary by CodeRabbit

* **New Features**
  * Added Python APIs for live-model training and prediction, MONAI Bundle interoperability, custom-model workflows, support-file packaging, and a two-class CPU segmentation quickstart.
  * Added resumable training, checkpoint caching, adaptive memory-bounded streaming, and performance benchmarking.
  * Added Studio lazy loading, retryable requests, improved viewer handling, accessible folder dialogs, and browser-based testing.

* **Bug Fixes**
  * Improved checkpoint selection, multipart uploads, job cancellation, cleanup, streamed patch handling, and stale viewer request handling.

* **Documentation**
  * Expanded guidance for custom models, large images, bundles, configuration, and performance.

### CODE REVIEW

File: benchmarks/perf/baselines/valentin-ThinkPad-P16-Gen-3.json [hunk 1]
```diff
@@ -0,0 +1,2537 @@
+{
+  "fingerprint": {
+    "date": "2026-09-07T17:23:41+0200",
+    "host": "valentin-ThinkPad-P16-Gen-3",
+    "git": {
+      "sha": "81a66be8db07",
+      "dirty": true,
```
[the remark this answers is not in the document]
Comment (vboussot) [reply benchmarks/perf/baselines/valentin-ThinkPad-P16-Gen-3.json:7]: The empty gate_warnings are by design: a series gates the load once, up front (KONFAI_PERF_GATED), and the load each later bench records is the series' own. The dirty flag is a real gap: the baseline is recaptured from a clean checkout on a quiet machine before the tag; this machine carries foreign work right now, so it stays open until then.

File: benchmarks/perf/bench_exaspim_budget.sh [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ROOT=$(cd "$HERE/../.." && pwd)
+OUT=$HERE/results/$(hostname)
+mkdir -p "$OUT"
+cd "$BENCH" || exit 1
+NAME=$(grep -oE '^\s*name: .*' Resample.yml | head -1 | awk '{print $2}')
+for B in "$@"; do
+  for REP in 1 2; do
+    cp Resample.yml Resample_sweep.yml
+    sed -i "s/memory_budget: .*/memory_budget: $B/" Resample_sweep.yml
+    rm -rf Registered "Transforms/$NAME"
```
[the remark this answers is not in the document]
Comment (vboussot) [reply benchmarks/perf/bench_exaspim_budget.sh:27]: Fixed in ba34fc34: an empty NAME refuses to clear Transforms/.

File: benchmarks/perf/bench_tests.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            summary = line.strip("= ")
+            break
+    counts = {"failed": 0, "passed": 0, "skipped": 0}
+    if summary:
+        for what in counts:
+            match = re.search(rf"(\d+) {what}", summary)
+            if match:
+                counts[what] = int(match.group(1))
+    # the -q summary line is not always the last line kept; the FAILED list is the ground truth
+    counts["failed"] = max(counts["failed"], len(failed))
```
[the remark this answers is not in the document]
Comment (vboussot) [reply benchmarks/perf/bench_tests.py]: Fixed in ba34fc34: pyproject addopts already carries -q, so the bench's own -q made -qq and silenced the summary line. Removed. No summary line is now an error, the FAILED list and the return code are the verdict, and _SUMMARY is gone.

File: konfai-apps/konfai_apps/bundle.py [hunk 4]
```diff
@@ -432,38 +432,95 @@ def assemble_bundle(
...
+        if target.resolve() in retained:
+            continue  # './model.pt' and 'model.pt' can name the same retained checkpoint.
+        if target.is_file():
+            obsolete.append(target)
+    for staged in staging.iterdir():
+        target = bundle / staged.name
+        if target.is_dir():
+            shutil.rmtree(target)
+        staged.replace(target)
+    staging.rmdir()
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai-apps/konfai_apps/bundle.py]: Addressed in dd704c9c: _replace_bundle materializes the staged listing (a sorted rglob) before any move, and replaces file by file.

File: konfai-mcp/konfai_mcp/server_jobs.py [hunk 5]
```diff
@@ -379,6 +381,30 @@ def _persist_manifest(
...
+        configuration that produced the run. The manifest names the copy under
+        ``resolved_config_snapshot``; a job that predates this records nothing."""
+        if self.workspace_layout is None or job.manifest_path is None or not job.manifest_path.exists():
+            return
+        try:
+            manifest = json.loads(job.manifest_path.read_text(encoding="utf-8"))
+        except (OSError, json.JSONDecodeError):
+            return
+        if "resolved_config_snapshot" in manifest or not job.config_path.exists():
+            return
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai-mcp/konfai_mcp/server_jobs.py:396]: Fixed in ba34fc34: the snapshot is taken on the transition to a terminal status only; a job already over when the server starts keeps what it has. Test added.

File: konfai/api.py [hunk 6]
```diff
@@ -463,16 +480,376 @@ def _config_copy(config: "Mapping[str, object] | Path | str") -> "dict[str, obje
...
+            f"No live model is registered under '{token}'.",
+            "A config naming 'konfai.api:live_model' runs in the process that built the model, through"
+            " konfai.train_model / konfai.predict_model; it cannot be replayed from its file alone.",
+        ) from None
+
+
+def _register_live_model(model: object) -> str:
+    token = f"{type(model).__name__}-{id(model):x}"
+    _LIVE_MODELS[token] = model
+    return token
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/api.py]: Fixed in ba34fc34: _registered_live_model is a context manager and the token is popped when the run returns. Test added.

File: konfai/utils/config.py [hunk 7]
```diff
@@ -610,6 +674,47 @@ def _parse_bool(value: object) -> bool:
...
+            raise ConfigError(f"'{where}' expects a list, got '{value}'.")
+        element_type = args[0] if args else Any
+        return [
+            _coerce_config_value(item, element_type, f"Element {index} of '{where}'")
+            for index, item in enumerate(value)
+        ]
+    if origin is dict:
+        if not isinstance(value, Mapping) or args[0] is not str or any(not isinstance(key, str) for key in value):
+            raise ConfigError(f"'{where}' expects a mapping with string keys, got '{value}'.")
+        return {key: _coerce_config_value(item, args[1], f"{where}.{key}") for key, item in value.items()}
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/config.py:713]: Fixed in ba34fc34: a tuple annotation binds a tuple, a fixed-length one coerces each position against its own annotation and refuses a wrong length. Tests added.

File: konfai/utils/dataset/h5.py [hunk 8]
```diff
@@ -338,6 +338,17 @@ def bounded_region_reads(self, name: str) -> bool:
...
+    def read_granularity(self, name: str) -> tuple[int, ...] | None:
+        """The chunk a chunked entry is stored in (``C[Z]YX``): a hyperslab decodes every chunk it
+        touches whole, so a window costs the chunk-aligned hull that covers it, and a sweep cut on
+        that grid reads each chunk once. ``None`` for a contiguous entry, what ``data_to_file``
+        writes, where a read costs the bytes it covers. A region write (``open_data_stream``)
+        chunks on its region, so a cache this run wrote answers the grain it was written in."""
+        dataset = cast(h5py.File, self.h5).get(name)
+        if not isinstance(dataset, h5py.Dataset) or dataset.chunks is None:
+            return None
+        return tuple(int(extent) for extent in dataset.chunks)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/utils/dataset/h5.py]: Fixed in ba34fc34: read_granularity resolves the entry through _get_dataset, so a recovered backup answers its chunk grain. Test added.

File: studio/frontend/tests/use-json.spec.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  const state = async () => JSON.parse(await page.locator("#json").innerText());
+
+  await page.evaluate(() => (window as any).__mount("JsonProbe", { url: "/api/evaluations?session=B" }));
+  await expect.poll(async () => (await state()).data?.runs?.[0]?.run).toBe("B");
+
+  // Switch to the slow session: B's rows must not stand in for A's while A loads.
+  await page.evaluate(() => (window as any).__mount("JsonProbe", { url: "/api/evaluations?session=A" }));
+  const during = await state();
+  expect(during.loading).toBe(true);
+  expect(during.data).toBeNull();
```
[the remark this answers is not in the document]
Comment (vboussot) [reply studio/frontend/tests/use-json.spec.ts]: Fixed in 024e027a: the test polls for the transient loading frame.

File: tests/unit/test_predictor_checkpoint_cache.py [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+    reads = _count_reads(composite, monkeypatch)
+    composite.load(paths)
+    assert composite._model_for_index(0).scale == 1
+    before = path.stat()
+    if replacement == "atomic":
+        temporary = tmp_path / "replacement.pt"
+        torch.save(_payload(9.0), temporary)
+        temporary.replace(path)
+    else:
+        torch.save(_payload(9.0), path)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply tests/unit/test_predictor_checkpoint_cache.py:149]: Fixed in 024e027a: mmap loads are off on win32, where a mapped file cannot be replaced; elsewhere the cache keeps mapped tensors and the test replaces the file as before.

File: tests/unit/test_trainer.py [hunk 11]
```diff
@@ -745,3 +748,336 @@ def test_a_score_that_is_not_finite_is_no_score(mode: str) -> None:
...
+    net.load(state, init=False, ema=False, override_lr=0.02)
+    assert scheduler.last_epoch == 0
+    assert net.optimizer.param_groups[0]["lr"] == 0.02
+
+
+def test_the_amp_scaler_state_rides_the_checkpoint() -> None:
+    net, _scheduler, ctx = _make_net(lambda opt: torch.optim.lr_scheduler.StepLR(opt, step_size=1, gamma=_GAMMA))
+    net.scaler = torch.amp.GradScaler("cuda", enabled=True, init_scale=4096.0)
+    states = net.schedule_states()
+    assert states["scaler"]["scale"] == 4096.0
```
[the remark this answers is not in the document]
Comment (vboussot) [reply tests/unit/test_trainer.py]: Fixed in afbc7e31: the test builds a CPU GradScaler, which stays enabled without CUDA.

File: benchmarks/perf/bench_tests.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+    after = resource.getrusage(resource.RUSAGE_CHILDREN)
+    cpu = (after.ru_utime - before.ru_utime) + (after.ru_stime - before.ru_stime)
+    text = _ANSI.sub("", completed.stdout + completed.stderr)
+    failed = sorted({line.split()[1] for line in text.splitlines() if line.startswith("FAILED ")})
+    summary = None
+    for line in reversed(text.splitlines()):
+        if " in " in line and ("passed" in line or "failed" in line):
+            summary = line.strip("= ")
+            break
+    if summary is None:
```
[the remark this answers is not in the document]
Comment (vboussot) [reply benchmarks/perf/bench_tests.py:69]: Fixed in 92ee2a41: a skipped-only summary is read; a line with none of passed/failed/skipped (no tests ran) is still an error.

File: examples/Segmentation/TwoClasses/quickstart.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+import json
+from pathlib import Path
+
+CASES = tuple(f"CASE_{index:03d}" for index in range(4))
+RUN = "CT_TWO_CLASSES"
+
+
+def prepare() -> None:
+    import numpy as np
+    import SimpleITK as sitk
```
[the remark this answers is not in the document]
Comment (vboussot) [reply examples/Segmentation/TwoClasses/quickstart.py]: Fixed in 92ee2a41: one guarded import helper used by prepare and verify, a handled error naming pip install 'konfai[itk]'.

File: konfai-apps/konfai_apps/bundle.py [hunk 14]
```diff
@@ -427,43 +498,137 @@ def assemble_bundle(
...
+    # Preflight the entire update: never write through a destination's symlink or replace a user's
+    # directory with a file. Updating a managed helper keeps unrelated files beside it.
+    for staged in staged_files:
+        target = bundle / staged.relative_to(staging)
+        for part in [target, *target.parents]:
+            if part == bundle:
+                break
+            if part.is_symlink():
+                raise AppMetadataError(f"Bundle destination contains a symlink: {part}")
+            if part != target and part.exists() and not part.is_dir():
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai-apps/konfai_apps/bundle.py]: Fixed in 92ee2a41: obsolete managed files are known before the preflight, a blocking ancestor among them is allowed, and they are removed before the moves. Test: a managed helper file becomes a support directory.

File: konfai/network/network/measure.py [hunk 15]
```diff
@@ -428,6 +449,53 @@ def set_window(self, n: int) -> None:
...
+        for group, records in self._loss.items():
+            saved = state["records"][group]
+            if saved.keys() != records.keys():
+                raise MeasureError("RESUME measure history does not match the configured criteria.")
+            for name, record in records.items():
+                entry = saved[name]
+                window = max(record._values.maxlen or 0, entry["window"] or len(entry["values"]), 1)
+                record._values = deque(entry["values"], maxlen=window)
+                record._weight = deque(entry["weights"], maxlen=window)
+                record._losses = deque(entry.get("losses", entry["values"]), maxlen=window)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/network/network/measure.py]: Fixed in 92ee2a41: a history without minimized losses restores none (empty deque, fresh running mean); the reported values are kept for the board and the selection starts from the epochs after the resume. Test added.

File: konfai-apps/konfai_apps/bundle.py [hunk 16]
```diff
@@ -427,43 +498,138 @@ def assemble_bundle(
...
+            if part == bundle:
+                break
+            if part.is_symlink():
+                raise AppMetadataError(f"Bundle destination contains a symlink: {part}")
+            if part != target and part.exists() and not part.is_dir() and part.resolve() not in removable:
+                raise AppMetadataError(f"Bundle destination parent is not a directory: {part}")
+        if target.is_dir():
+            raise AppMetadataError(f"Bundle destination is a directory, not a file: {target}")
+    for target in obsolete:
+        target.unlink()
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai-apps/konfai_apps/bundle.py:625]: Fixed in b6a5d43d: obsolete files are keyed by resolved path, so two spellings unlink once. Test added.
